### PR TITLE
Hotfix unique id after regeneration

### LINKED ISSUE #1885: References to PROTO regenerated nodes not updated in Supervisor API
After regenerating a PROTO node, the reference in the Supervisor API doesn't seem to be updated and `wb_supervisor_node_get_from_def` still returns the previous (before PROTO regeneration) node reference.

World file:
```
#VRML_SIM R2020b utf8
WorldInfo {
}
Viewpoint {
}
RectangleArena {
  floorAppearance PBRAppearance {
    baseColorMap DEF TEST_NODE ImageTexture {
      url [
        "textures/parquetry/chequered_parquetry_base_color.jpg"
      ]
    }
  }
}
Robot {
  controller "modify_proto_template_field"
  supervisor TRUE
}
```

C Controller:
```
#include <stdio.h>
#include <webots/robot.h>
#include <webots/supervisor.h>

#define TIME_STEP 32

int main(int argc, char **argv) {
  wb_robot_init();
  WbNodeRef node = wb_supervisor_node_get_from_def("TEST_NODE");
  printf("Node before: %p \n", node);
  WbFieldRef filteringField = wb_supervisor_node_get_field(node, "filtering");
  
  // If the url field reference is requested before regeneration,
  // then it is cached and returned after regeneration even if invalid
  //WbFieldRef urlFieldBefore = wb_supervisor_node_get_field(node, "url");
  //printf("Url field before: %p \n", urlFieldBefore);
  

  wb_robot_step(TIME_STEP);
  
  // trigger RectangleArena regeneration
  wb_supervisor_field_set_sf_int32(filteringField, 2);
  
  wb_robot_step(TIME_STEP);
  
  node = wb_supervisor_node_get_from_def("TEST_NODE");
  printf("Node after: %p \n", node);
  // Invalid node reference returned and url field is not found 
  WbFieldRef urlField = wb_supervisor_node_get_field(node, "url");
  printf("Url field after: %p \n", urlField);

  wb_robot_cleanup();
  return 0;
}
```

### PR DESCRIPTION
**Pull request by stefaniapedrazzi**
Fix #1885.

Unique ids of PROTO parameter nodes are not correctly restored after procedural PROTO regeneration.
This causes issues in the controller library where it is assumed that the ids remains the same and it is no longer possible to access the nodes.
(No notification is received in the libController that nodes have been deleted during regeneration and thus it is also not possible to retrieve the new nodes with the new unique ids).

I fixed the unique ids for the PROTO parameters but not for internal PROTO nodes (the ones retrieved by `wb_supervisor_node_get_from_proto_def`). This case is much more complicated and it is even not ensured that the node structure remain the same.

A second issues was caused by notifying the deletion of regenerated PROTO nodes to the supervisor controller in two steps:
1. add node to the WbSupervisorUtilities::mNodesDeletedSinceLastStep
2. propagate the WbSupervisorUtilities::mNodesDeletedSinceLastStep to the controller

At step 1), the node reference was still valid, but the same was not guaranteed at step 2).
Checking if deleted nodes are references by other nodes fixes the issues and prevents to use invalid pointers.

### CODE REVIEW

**Comment by stefaniapedrazzi** [comment]
It is not possible to guarantee the validity of internal PROTO fields and nodes after PROTO regeneration on the libController side.
Even detecting exactly which node and field became invalid could be complicated.
Instead, I introduced a message that will invalidate all the internal PROTO fields and nodes after PROTO regeneration without checking if they were descendant of the regenerated node or not.

The libController is now able to detect these invalidated references and returns an error so that the user can retrieve the new reference.

**Comment by omichel** [comment]
:eyes:

File: src/lib/Controller/api/supervisor.c [hunk 1]
```diff
@@ -235,6 +238,46 @@ static const char *extract_node_def(const char *def_name_expression) {
...
+  while (field) {
+    if (field->is_proto_internal) {
+      if (previous_field)
+        previous_field->next = field->next;
+      else
+        field_list = field->next;
+      WbFieldStruct *current_field = field;
+      field = field->next;
+      // clean the field
+      printf("remove_internal field %p\n", current_field);
```
Comment (omichel) [inline src/lib/Controller/api/supervisor.c:270]: We should probably remove this debug statement.
```suggestion
```

**Comment by stefaniapedrazzi** [comment]
Thank you for the review.

I just added a new condition to avoid that an internal proto node could be returned instead of a visible DEF node if sharing the same DEF name.

**Pull request merged**